runtime/vm/exceptiondescribe.c

```diff
@@ -275,30 +301,86 @@ iterateStackTrace(J9VMThread * vmThread, j9object_t* exception, callback_func_t
 		while (currentElement != arraySize) {
 			/* write as for or move currentElement++ to very end */ 
 			UDATA methodPC = J9JAVAARRAYOFUDATA_LOAD(vmThread, J9VMJAVALANGTHROWABLE_WALKBACK(vmThread, (*exception)), currentElement);
+			/* The methodPC value is overwritten by the bytecode index if the methodPC points to a JIT frame, but we want
+			 * to cache the original address hence we need to save a copy here.
+			 */
+			UDATA methodPCOriginal = methodPC;
+			BOOLEAN methodPCCacheHit = FALSE;
 			J9ROMMethod * romMethod = NULL;
 			J9ROMClass *romClass = NULL;
 			UDATA lineNumber = 0;
 			J9UTF8 * fileName = NULL;
 			J9ClassLoader *classLoader = NULL;
+
 #ifdef J9VM_INTERP_NATIVE_SUPPORT
 			J9JITExceptionTable * metaData = NULL;
 			UDATA inlineDepth = 0;
+			/* The inlineDepth value is mutated in the loops below as we walk through the inlining table. We are only
+			 * chaching the toplevel JIT frame so we need to save a copy of the original value.
+			 */
+			UDATA inlineDepthOriginal = 0;
 			void * inlinedCallSite = NULL;
 			void * inlineMap = NULL;
 			J9JITConfig * jitConfig = vm->jitConfig;
+			MethodPCCache *methodPCCache = vm->methodPCCache;
+			MethodPCCache *cacheEntry = NULL;
+
+			if (NULL == methodPCCache) {
+				MethodPCCache *existingCache = NULL;
+				PORT_ACCESS_FROM_JAVAVM(vm);
+				methodPCCache = j9mem_allocate_memory(METHODPC_CACHE_SIZE * sizeof(MethodPCCache), OMRMEM_CATEGORY_JIT);
+
+				if (NULL != methodPCCache) {
+					memset(methodPCCache, 0, METHODPC_CACHE_SIZE * sizeof(MethodPCCache));
+
+					/* The vmThread parameter to this function may not be the current thread, so ensure that only a single
+					 * instance of the cache is allocated for the VM, and make sure the empty cache entries for a new cache
+					 * are visible to other threads before the cache pointer is visible.
+					 */
+					issueWriteBarrier();
+
+					existingCache = (MethodPCCache*)compareAndSwapUDATA((uintptr_t*)&vm->methodPCCache, (uintptr_t)existingCache, (uintptr_t)methodPCCache);
+					if (NULL != existingCache) {
+						j9mem_free_memory(methodPCCache);
+						methodPCCache = existingCache;
+					}
+				}
+			}
```

Comment:
Definitely not the best place to do this. The best place I was able to find was in `protectedInitializeJavaVM` in jvminit.c, however I'm not sure where in in there to place this code. Placing it in such a location would require constants such as `METHODPC_CACHE_SIZE` to be globally exposed. Not sure where I would put them either. Any advice here from reviewers would be appreciated.